wfe/probs.go

```diff
@@ -0,0 +1,79 @@
+package wfe
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/letsencrypt/boulder/core"
+	berrors "github.com/letsencrypt/boulder/errors"
+	"github.com/letsencrypt/boulder/probs"
+)
+
+// problemDetailsForError turns an error into a ProblemDetails with the special
+// case of returning the same error back if its already a ProblemDetails. If the
+// error is of an type unknown to ProblemDetailsForError, it will return a
+// ServerInternal ProblemDetails.
+//
+// TODO(XXX): These translations should be programmatically generated in the boulder/errors
+// package so we have everything in one place
+func problemDetailsForError(err error, msg string) *probs.ProblemDetails {
+	switch e := err.(type) {
+	case *probs.ProblemDetails:
+		return e
+	case *berrors.BoulderError:
+		switch e.Type {
+		case berrors.NotSupported:
+			return &probs.ProblemDetails{
+				Type:       probs.ServerInternalProblem,
+				Detail:     fmt.Sprintf("%s :: %s", msg, err),
+				HTTPStatus: http.StatusNotImplemented,
+			}
+		case berrors.Malformed, berrors.SignatureValidation:
+			return probs.Malformed(fmt.Sprintf("%s :: %s", msg, err))
+		case berrors.Unauthorized:
+			return probs.Unauthorized(fmt.Sprintf("%s :: %s", msg, err))
+		case berrors.NotFound:
+			return probs.NotFound(fmt.Sprintf("%s :: %s", msg, err))
+		case berrors.RateLimit:
+			return probs.RateLimited(fmt.Sprintf("%s :: %s", msg, err))
+		case berrors.InternalServer, berrors.TooManyRequests:
+			// Internal server error messages may include sensitive data, so we do
```

Comment:
This is different than the semantics we've been using for InternalServerError, where is some component in an RPC call chain specifically creates an InternalServerError (as opposed to errors.New), that is a way of saying "the detail I'm providing is fine to pass up to the user." The code here is saying, essentially, that only the WFE is allowed to set the detail for an InternalServerError.